rbreslow/zs
# Add an administrative system

## Overview

Includes command parsing, a permissions system, a process for saving restoring a player's profile, checking for bans when a player connects.

Resolves #11 

## Testing Instructions

Run `scripts/update` and see that migrations are applied.

```console
$ ./scripts/update
. . .
SQL: SELECT COUNT(*) as c from pg_class where relname = 'lapis_migrations'
SQL: SELECT * from "lapis_migrations"
Migrating: 1599790650
SQL: CREATE TABLE "player" (
  "id" bigint,
  "name" character varying(255),
  "role" character varying(255) NOT NULL DEFAULT 'user',
  "updated_at" timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP,
  PRIMARY KEY ("id")
);
SQL: CREATE INDEX "player_name_idx" ON "player" ("name");
SQL: INSERT INTO "player" ("id", "role", "name") VALUES (-1, '', 'Console')
SQL: CREATE TABLE "ban" (
  "id" serial NOT NULL PRIMARY KEY,
  "player_id" bigint,
  "banned_by" bigint,
  "banned_at" timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP,
  "expires_at" timestamp NOT NULL,
  "reason" character varying(255),
  "unbanned" boolean NOT NULL DEFAULT FALSE,
  FOREIGN KEY ("player_id") REFERENCES "public"."player"("id"),
  FOREIGN KEY ("banned_by") REFERENCES "public"."player"("id")
);
SQL: CREATE INDEX "ban_player_id_expires_at_unbanned_idx" ON "ban" ("player_id", "expires_at", "unbanned");
SQL: CREATE INDEX "maps_last_played_installed_idx" ON "maps" ("last_played", "installed");
SQL: INSERT INTO "lapis_migrations" ("name") VALUES ('1599790650') RETURNING "name"
Ran 1 migration
```

See that you can ban a user by their Steam ID and they will fail to connect:

```console
howl banid STEAM_0:0:11101 "30 seconds" test
[2020-10-19 01:44:56 +0000] [INFO] gamemodes/zombiesurvival/gamemode/howl/sv_howl.lua:136: Console (-1) executed banid STEAM_0:0:11101 "30 seconds" test.
Console banned 76561197960287930 for 30 seconds ('test')
[2020-10-19 01:45:00 +0000] [INFO] gamemodes/zombiesurvival/gamemode/howl/sv_hooks.lua:31: Stereo Love (172.21.0.1:40194) attempted to connect. Disconnect: You are banned for 26 seconds. Reason: test.
```

See that you can unban a user with their Steam ID:

```console
howl banid STEAM_0:0:11101 "30 seconds" test
[2020-10-19 01:48:05 +0000] [INFO] gamemodes/zombiesurvival/gamemode/howl/sv_howl.lua:136: Console (-1) executed banid STEAM_0:0:11101 "30 seconds" test.
Console banned 76561197960287930 for 30 seconds ('test')
```
```console
howl unban STEAM_0:0:11101
[2020-10-19 01:48:08 +0000] [INFO] gamemodes/zombiesurvival/gamemode/howl/sv_howl.lua:136: Console (-1) executed unban STEAM_0:0:11101.
Unbanning 76561197960287930...
Stereo Love (76561197960287930) has been unbanned.
```
```console
howl unban STEAM_0:0:11101
[2020-10-19 01:48:12 +0000] [INFO] gamemodes/zombiesurvival/gamemode/howl/sv_howl.lua:136: Console (-1) executed unban STEAM_0:0:11101.
Unbanning 76561197960287930...
No bans found for 76561197960287930.
```

Make yourself an admin:

```console
howl setrole stereo admin
[2020-10-19 01:51:13 +0000] [INFO] gamemodes/zombiesurvival/gamemode/howl/sv_howl.lua:136: Console (-1) executed setrole stereo admin.
Set Stereo Love's role to admin.
```

From Garry's Mod, see that you have access to the full suite of commands:

![image](https://user-images.githubusercontent.com/1774125/96393076-23e16600-118c-11eb-8594-ef7af03c2d1a.png)

## Checklist
